core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java

```diff
@@ -125,8 +126,39 @@ protected final void updateTopologyId(TopologyAffectedCommand command) throws In
       // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
       int newTopologyId = Math.max(currentTopologyId(), commandTopologyId + 1);
       command.setTopologyId(newTopologyId);
-      waitForTransactionData(newTopologyId);
-      return visitGetKeysInGroupCommand(context, command);
+      return retryCommandWithTransactionData(newTopologyId, context, command, (ctx, cmd) -> visitGetKeysInGroupCommand(ctx, cmd));
+
+   }
+
+   protected <T extends VisitableCommand> CompletableFuture<Void> retryCommandWithTransactionData(int topologyId,
+        InvocationContext ctx, T command, RetryHandler<T> handler) throws Throwable {
+      CompletableFuture<Void> future = stateTransferLock.transactionDataFuture(topologyId);
+      if (future == null) {
+         getLog().tracef("Retrying command %s for topology %d", command, topologyId);
+         return handler.handle(ctx, command);
+      } else {
+         CompletableFuture<Void> retryFuture = future.thenCompose(ignored -> {
+            try {
+               getLog().tracef("Retrying command %s for topology %d", command, topologyId);
+               return handler.handle(ctx, command);
+            } catch (Throwable t) {
+               throw new CacheException(t);
+            }
+         });
+         // We want to time out the current command future, not the main topology-waiting future,
+         // but the command future can take longer time to finish.
+         ScheduledFuture<?> timeoutFuture = timeoutExecutor.schedule(() -> {
+            if (!future.isDone()) {
+               retryFuture.completeExceptionally(new TimeoutException("Timed out waiting for topology " + topologyId));
```

Comment:
Right, I forgot that... Cancelling that future would have been ok if the timeout was handled in `StateTransferLockImpl`, because all the futures were using the same timeout, but here we don't have any guarantees that other commands that don't go through `StateTransferInterceptor` use the same timeout. 

Still, I don't think completing `retryFuture` will stop the handler from running, and that can be a problem.